include/pybind11/detail/common.h

```diff
@@ -306,13 +306,21 @@ extern "C" {
             });
         }
 \endrst */
+#if PY_MAJOR_VERSION >= 3
+#define PYBIND11_MODULE_PY2_VS_PY3_SPECIFIC(name)                              \
+        static PyModuleDef mdef;                                               \
+        auto m = pybind11::module(PYBIND11_TOSTRING(name), nullptr, &mdef);
+#else
+#define PYBIND11_MODULE_PY2_VS_PY3_SPECIFIC(name)                              \
+        auto m = pybind11::module(PYBIND11_TOSTRING(name));
+#endif
 #define PYBIND11_MODULE(name, variable)                                        \
     PYBIND11_MAYBE_UNUSED                                                      \
     static void PYBIND11_CONCAT(pybind11_init_, name)(pybind11::module &);     \
     PYBIND11_PLUGIN_IMPL(name) {                                               \
         PYBIND11_CHECK_PYTHON_VERSION                                          \
         PYBIND11_ENSURE_INTERNALS_READY                                        \
-        auto m = pybind11::module(PYBIND11_TOSTRING(name));                    \
+        PYBIND11_MODULE_PY2_VS_PY3_SPECIFIC(name)                              \
```

Comment:
Woudn't this be better _not_ factored out into a macro when we remove Python 2 support? Then you would just drop the `#else` part and be done with it, rather than now having to remove the extra macro.

Reply:
> Woudn't this be better not factored out into a macro when we remove Python 2 support?

Thoughts regarding duplicating the entire PBBIND11_MODULE macro:

* Very slightly easier to purge Python 2 support. However, even someone who's never seen this code before can purge Python 2 support from the code as-is in less than 5 minutes (just move two lines down).

* Code is read more often than it is written. Assume 100 people look at the 11 (out of 12 or 13) lines copied here. It will take most people at least a couple minutes to conclusively determine that the 11 lines are exactly identical. So the total time spent will be 200 minutes.

5 minutes << 200 minutes.
That doesn't even take into account time lost to oversights when this code needs to be modified in two places.

I feel strongly not duplicating this code is the right call here.

Reply:
It's a long macro, yes. Maybe the problem is the name that makes this look weird, @henryiii, since it containss PY2/PY3?
Why don't we just call this `PYBIND11_CREATE_MODULE` (which is what it does, so it's easier to read for those 100 people reading this part of the code without looking at the macro above). Then forgetting to remove the macro when dropping Python 2 wouldn't even be an issue, since it does still create a module.

Reply:
I guess the problem is that you can't have an `#if` statement in a macro. Yes, then @YannickJadoul solution is good, as it describes what does. Since macros are global, having a relatively easy and not clearly "detail" or "internal" name might be an issue?

Reply:
Ah, yes, the main problem with macros. If it's not documented, then using it is at the user's own risk, though?

Reply:
Like `PYBIND11_NAMESPACE`?

Reply:
> I guess the problem is that you can't have an `#if` statement in a macro. Yes, then @YannickJadoul solution is good, as it describes what does. Since macros are global, having a relatively easy and not clearly "detail" or "internal" name might be an issue?

Thanks @YannickJadoul and @henryiii , I changed the helper macro name to `PYBIND11_MODULE_DETAIL_CREATE`, taking ideas from both of you.

Generally, I prefer very long macro names for private macros ("privacy by intentional obscurity"), e.g. this could be `PYBIND11_MODULE_DETAIL_PY2_VS_PY3_SPECIFIC`, but I get the sense shorter names are preferred here.

Reply:
We should try figuring out a general approach, at some point. Is there a way projects tend to handle "private macros" (unfortunately, leading underscores are reserved in C++) ? (Curious, though: why is `DETAIL` inserted right in the very middle of the name; wouldn't `PYBIND11_DETAIL` make more sense as an emerging convention, similar to `pybind11::detail`?)

Then again, not documenting is good enough. I'm really not going to lose sleep over users using a non-documented macro; they shouldn't, and if they do, they should know what the risks are?

Reply:
> `PYBIND11_MODULE_DETAIL_PY2_VS_PY3_SPECIFIC`, but I get the sense shorter names are preferred here.

PS: My problem wasn't per se the long name (though not a fan, and I doubt if long, descriptive names will stop users for accessing it), but the thing that the name specifies the implementation. Suppose e.g. PyPy has a problem and needs a different way of handling the creation of a module, then you'd need to change the name; now you don't.